### DISCUSSION TITLE
Any examples of `lsp-mode` and Meow?

### DISCUSSION THREAD

**Discussion by zetashift** [meow-edit/meow#392] [Q&A] [answered]
I looked around for a bit and I was wondering if there are any configs out there that show meow working with lsp-mode. Preferably using the `space` as the leader key, and `g` to do some `go-to` actions.

As I'm a beginner with Meow I'm not entirely sure how to do the mappings only while lsp-mode is active.

**Comment by DogLooksGood** [comment, marked as answer]
Hey, here is my configuration about `lsp-mode`. I didn't remap actually, just use `SPC c l`.

https://github.com/DogLooksGood/meomacs/blob/master/programming.org#lsp-mode

The default leader dispatch is `mode-specific-map`. To have `SPC l` for `C-c l`, you can add this
```
(setq meow-keypad-leader-dispatch "C-c")
```

**Reply by zetashift** [reply]
Oh very nice! Thank you :D!

**Reply by zetashift** [reply]
Where do I add the `(setq meow-keypad-leader-dispatch "C-c")` part? I added it in the `:config` of my `use-package meow` form, but pressing `SPC l` for example doesn't bring up lsp keybindings or `SPC b b` doesn't list my buffers in perspective.el

**Reply by DogLooksGood** [reply]
I think you can add it anywhere you want. It's just an variable.  Are you using Doom Emacs?
If you are using Doom Emacs, you probably want to set it to `doom-leader-map`.

**Comment by zetashift** [comment]
I'm having trouble as well with writing good elisp and my keybindings.

For global operations using the leader key, I currently have the following elisp:

```elisp
  (meow-leader-define-key
   ;; SPC j/k will run the original command in MOTION state.
   '("j" . "H-j")
   '("k" . "H-k")
   '("fs" . save-buffer)
   '("ff" . find-file)
   '("bb" . switch-to-buffer)
   '("bs" . scratch-buffer)
   '("bd" . kill-this-buffer)
   '("wv" . split-window-vertically)
   '("ws" . split-window-horizontally)
   '("ww" . next-window-any-frame)
   '("wq" . delete-window)
   '("eb" . eval-buffer)
   '("qr" . restart-emacs)
;; etc
```

But how do I make this more elegant, and make it work with which-key? Define a keymap? but then I don't know how to use it with which-key still :(.

**Reply by eshrh** [reply]
What exactly doesn't work with which-key here? Seems to work how I expect. Pressing SPC-f shows me both s for save-buffer and f for find-file.

I'm a bit confused about what you want to improve, it looks fine to me.

**Reply by zetashift** [reply]
So for example I have `SPC f s` for `save-buffer` and `SPC f f` for `find-file`.
If I press "SPC", which-key pops up and guides me with `f - prefix`. I would like to replace that "prefix" label with "file"
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/1857826/217701980-5f18b84d-4e91-4f79-a5e4-671ec3df2650.png">

which-key has commands for that:
```
(which-key-add-keymap-based-replacements some-map
  "b" '("bar-prefix" . (keymap)))
```
And some others. But I don't really know how to use them together with `meow-leader-define-key`

**Reply by DogLooksGood** [reply]
meow can read those built-in keymap names because they have alias name.

In emacs source code, there's something like this. So meow will know the keymap's name.
```
(defalias 'kmacro-keymap kmacro-keymap)
```

But in which-key, keymap names are not bound to keymap themselves, it's bound to the key binding.  So some settings won't work.

**Reply by zetashift** [reply]
So does that mean there is no way for me to replace those `+prefix`labels for `f` or `b` keymaps?
Sorry for the dumb questions, usage wise, I fixed the LSP problems and I'm having a blast with meow! Just have this little wish to neatly label everything and not have `+prefix` littered all over.

**Reply by DogLooksGood** [reply]
Only if you name these keymap with aliases manually, like how Emacs built-in keymaps are handled.  I haven't found a way to workaround this yet. 

If we can rebuild the KEYPAD state with real key translation, then it's possible.

**Reply by zetashift** [reply]
So I'd have to do something like

`(defalias 'filekeymap file)`?

Anyhow thank you for your time!
